'digits' tolType

Alex-Jordan opened the pull request:
This adds `tolType=>'digits'` as an option for a context flag. Documentation will be required, but I am not 100% sure of design decisions I made here. This PR may lead to discussion that changes the design, and then I can make documentation.

As implemented right now, when this is set, the `tolerance` flag should be a natural number 1, 2, 3, etc. Roughly, the idea is that if two numbers "match" for their first so many significant digits, then the `compare` subroutine will see no difference between them. (And if `tolerance` is something like 0.001, the default with relative tolerance, then that gets converted to 3 automatically using log10.)

It doesn't distinguish things like 1.2 and 1.200, so I decided not to call this a `sigfig` tolType.

If `tolerance` is 3 and the answer is pi, should 3.14177 count as equal? I've decided it should not, even though it matches on the first 3 digits. This kind of extra checking is done up to the 6th significant digit, where 6 is a new context flag `tolDigits`. 6 is the default, because that is the default number of digits MathObjects show (because the number format is `%g`, which uses 6 as its default.)

If `tolerance` is 4 and the answer is pi, should 3.141 count as equal? Because rounding pi would give 3.142. So I added another flag `tolTruncation` that tolerates truncation. It is true by default, and both 3.141 and 3.142 would be equal to pi. If it's set to false, only 3.142 is equal to pi.

Below is a PG file I used to do some testing.


```
DOCUMENT();

loadMacros(
  "PGstandard.pl",
  "MathObjects.pl",  
  "PGML.pl",
  "niceTables.pl",
  "PGcourse.pl",
);

TEXT(beginproblem());

Context("Numeric")->flags->set(tolType=>'digits',tolerateTruncation=>0);
# Note the default tolerance of 0.001 is converted to tolerance of 3

@a = (
Real(pi),
Real(3.14),
Real(3.15),
Real(3.13),
Real(3.141),
Real(3.142),
Real(3.14159),
Real(3.14149),
Real(0.1055),
Real(0.105),
Real(0.106),
Real(0.1155),
Real(0.115),
Real(0.116),
Real(-0.1055),
Real(-0.105),
Real(-0.106),
Real(-0.1155),
Real(-0.115),
Real(-0.116),
Real(1234e20),
Real(123000000000000000000000),
);

BEGIN_PGML
With 'digits' tolerance, with 'tolerance' 3, and 'tolTruncation' true.
[@ DataTable(
  [
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[1] . '\)',
      ($a[0] == $a[1]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[2] . '\)',
      ($a[0] == $a[2]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[3] . '\)',
      ($a[0] == $a[3]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[4] . '\)',
      ($a[0] == $a[4]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[5] . '\)',
      ($a[0] == $a[5]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[6] . '\)',
      ($a[0] == $a[6]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[7] . '\)',
      ($a[0] == $a[7]) ? 'True' : 'False'],
    ['\(' . $a[8] . '\stackrel{?}{=}' . $a[9] . '\)',
      ($a[8] == $a[9]) ? 'True' : 'False'],
    ['\(' . $a[8] . '\stackrel{?}{=}' . $a[10] . '\)',
      ($a[8] == $a[10]) ? 'True' : 'False'],
    ['\(' . $a[11] . '\stackrel{?}{=}' . $a[12] . '\)',
      ($a[11] == $a[12]) ? 'True' : 'False'],
    ['\(' . $a[11] . '\stackrel{?}{=}' . $a[13] . '\)',
      ($a[11] == $a[13]) ? 'True' : 'False'],
    ['\(' . $a[14] . '\stackrel{?}{=}' . $a[15] . '\)',
      ($a[14] == $a[15]) ? 'True' : 'False'],
    ['\(' . $a[14] . '\stackrel{?}{=}' . $a[16] . '\)',
      ($a[14] == $a[16]) ? 'True' : 'False'],
    ['\(' . $a[17] . '\stackrel{?}{=}' . $a[18] . '\)',
      ($a[17] == $a[18]) ? 'True' : 'False'],
    ['\(' . $a[17] . '\stackrel{?}{=}' . $a[19] . '\)',
      ($a[17] == $a[19]) ? 'True' : 'False'],
    ['\(' . $a[20] . '\stackrel{?}{=}' . $a[21] . '\)',
      ($a[20] == $a[21]) ? 'True' : 'False'],
  ],
  center=>0,
  align=>'ll',
) @]*
----
END_PGML

Context()->flags->set(tolTruncation=>0);

BEGIN_PGML
Now with 'tolTruncation' false.
[@ DataTable(
  [
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[1] . '\)',
      ($a[0] == $a[1]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[2] . '\)',
      ($a[0] == $a[2]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[3] . '\)',
      ($a[0] == $a[3]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[4] . '\)',
      ($a[0] == $a[4]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[5] . '\)',
      ($a[0] == $a[5]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[6] . '\)',
      ($a[0] == $a[6]) ? 'True' : 'False'],
    ['\(' . $a[0] . '\stackrel{?}{=}' . $a[7] . '\)',
      ($a[0] == $a[7]) ? 'True' : 'False'],
    ['\(' . $a[8] . '\stackrel{?}{=}' . $a[9] . '\)',
      ($a[8] == $a[9]) ? 'True' : 'False'],
    ['\(' . $a[8] . '\stackrel{?}{=}' . $a[10] . '\)',
      ($a[8] == $a[10]) ? 'True' : 'False'],
    ['\(' . $a[11] . '\stackrel{?}{=}' . $a[12] . '\)',
      ($a[11] == $a[12]) ? 'True' : 'False'],
    ['\(' . $a[11] . '\stackrel{?}{=}' . $a[13] . '\)',
      ($a[11] == $a[13]) ? 'True' : 'False'],
    ['\(' . $a[14] . '\stackrel{?}{=}' . $a[15] . '\)',
      ($a[14] == $a[15]) ? 'True' : 'False'],
    ['\(' . $a[14] . '\stackrel{?}{=}' . $a[16] . '\)',
      ($a[14] == $a[16]) ? 'True' : 'False'],
    ['\(' . $a[17] . '\stackrel{?}{=}' . $a[18] . '\)',
      ($a[17] == $a[18]) ? 'True' : 'False'],
    ['\(' . $a[17] . '\stackrel{?}{=}' . $a[19] . '\)',
      ($a[17] == $a[19]) ? 'True' : 'False'],
    ['\(' . $a[20] . '\stackrel{?}{=}' . $a[21] . '\)',
      ($a[20] == $a[21]) ? 'True' : 'False'],
  ],
  center=>0,
  align=>'ll',
) @]*
END_PGML

ENDDOCUMENT();

```

Alex-Jordan commented:
I think this will be a valuable tolerance alternative to `relative` and `absolute`, especially at lower levels. We've used this at my school. A lot actually, since we additionally changed the default tolerance type of Numeric to be `digits` instead of `relative`. And then that leads to most contexts inheriting `digits` as the tolType. It has worked as expected.

My sample problem has a mistake. It has `tolerateTruncation=>0` that should be `tolTruncation=>1`. The mistake is inconsequential, since the default for `tolTruncation` is 1.

Alex-Jordan commented:
See https://webwork.maa.org/moodle/mod/forum/discuss.php?d=5004#p15278.

I need to adjust this to handle when one or both numbers are within the $zeroLevel .

Alex-Jordan commented:
Here is a cleaned up test file, which includes tests with numbers close to 0.

```
DOCUMENT();

loadMacros(
  "PGstandard.pl",
  "MathObjects.pl",  
  "PGML.pl",
  "niceTables.pl",
  "PGcourse.pl",
);

TEXT(beginproblem());

@pi = (
Real(pi),
Real(3.14),
Real(3.15),
Real(3.13),
Real(3.141),
Real(3.142),
Real(3.14159),
Real(3.14149),
);

@lz = (
Real(0.1055),
Real(0.105),
Real(0.106)
);

@nlz = (
Real(-0.1055),
Real(-0.105),
Real(-0.106)
);

@lg = (
Real(1234e20),
Real(123000000000000000000000),
Real(124000000000000000000000),
);

@sm = (
Real(0),
Real(10**(-14)),
Real(10**(-15)),
);

$TBL = <<'END';
[@ DataTable(
  [
    (map{['\('.$pi[0].'\stackrel{?}{=}'.$_.'\)',($pi[0]==$_)?'T':'F']}(@pi)),
    (map{['\('.$lz[0].'\stackrel{?}{=}'.$_.'\)',($lz[0]==$_)?'T':'F']}(@lz)),
    (map{['\('.$nlz[0].'\stackrel{?}{=}'.$_.'\)',($nlz[0]==$_)?'T':'F']}(@nlz)),
    (map{['\('.$lg[0].'\stackrel{?}{=}'.$_.'\)',($lg[0]==$_)?'T':'F']}(@lg)),
    (map{['\('.$sm[0].'\stackrel{?}{=}'.$_.'\)',($sm[0]==$_)?'T':'F']}(@sm)),
  ],
  center=>0,
  align=>'ll',
) @]*
END

Context("Numeric")->flags->set(tolType=>'digits',tolTruncation=>1);
# Note the default tolerance of 0.001 is converted to tolerance of 3

BEGIN_PGML
With 'digits' tolerance, with 'tolerance' 3, and 'tolTruncation' true.  
[$TBL]**
----
END_PGML

Context()->flags->set(tolTruncation=>0);

BEGIN_PGML
Now with 'tolTruncation' false.  
[$TBL]**
END_PGML

ENDDOCUMENT();
```

drgrice1 reviewed (approved):
This looks okay to me.  It only does anything if the tolType flag is the new 'digits' value, so it will not affect existing problems.

One minor clean up issue though.

Alex-Jordan commented on lib/Value/Real.pm:
```diff
@@ -130,6 +130,30 @@ sub compare {
+      return 0;
+    }
     if ($self->getFlag('tolType') eq 'relative') {
       my $zeroLevel = $self->getFlag('zeroLevel');
```
Nice catch, I added a commit to remove the redundant line.

drgrice1 commented:
Could someone else take a look at this.  This is ready to be merged in I think.

drdrew42 commented:
With the updated sample problem, I'm not seeing any different results between tolTruncation being true or false.

Alex-Jordan commented:
I also see no difference now. With tolTruncation false, pi should not count as equal to 3.141. But they should be equal with tolTrucnation true...

 I will invesitgate.

drgrice1 commented:
I didn't try testing with your later version.  I see that with the original test file you posted there still is a difference.  But with the newer one, there is not.

drgrice1 commented:
One thing that might help for those of us testing this is if you added a column to the output of the test file with the expected result.

Alex-Jordan commented:
I think it has to do with how the new test file builds the table before context settings are finalized. New test file coming shortly.

Alex-Jordan commented:
Here is a new test problem. First there is a table for when tolTruncation is true. Then it makes a new context, setting tolTruncation to false, and rebuilds everything in that context. The table construction is to make this sample problem code be as streamlined as I can easily make it. But if I don't just rebuild the table completely on a second run, then I think the context is keeping comparison results in some cases.

In the second table, things should change for 3.141, 0.1055, and -0.1055.

```
DOCUMENT();

loadMacros(
  "PGstandard.pl",
  "MathObjects.pl",  
  "PGML.pl",
  "niceTables.pl",
  "PGcourse.pl",
);

TEXT(beginproblem());

BEGIN_PGML
With 'digits' tolerance, with 'tolerance' 3, and 'tolTruncation' true.  
END_PGML

Context("Numeric")->flags->set(tolType=>'digits',tolTruncation=>1);
# Note the default tolerance of 0.001 is converted to tolerance of 3

# a value, then the expecations for comparison to 0th value
@pi = (
[Real(pi),'T','T'],
[Real(3.14),'T','T'],
[Real(3.15),'F','F'],
[Real(3.13),'F','F'],
[Real(3.141),'T','F'],
[Real(3.142),'T','T'],
[Real(3.14159),'T','T'],
[Real(3.14149),'F','F'],
);

@lz = (
[Real(0.1055),'T','T'],
[Real(0.105),'T','F'],
[Real(0.106),'T','T'],
);

@nlz = (
[Real(-0.1055),'T','T'],
[Real(-0.105),'T','F'],
[Real(-0.106),'T','T'],
);

@lg = (
[Real(1234e20),'T','T'],
[Real(123000000000000000000000),'T','T'],
[Real(124000000000000000000000),'F','F'],
);

@sm = (
[Real(0),'T','T'],
[Real(10**(-14)),'F','F'],
[Real(10**(-15)),'T','T'],
);


$TBL = <<'END';
[@ DataTable(
  [
    ['','Equal?','Expecation allowing truncation','Expecation not allowing truncation'],
    (map{['\('.$pi[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($pi[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@pi)),
    (map{['\('.$lz[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($lz[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@lz)),
    (map{['\('.$nlz[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($nlz[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@nlz)),
    (map{['\('.$lg[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($lg[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@lg)),
    (map{['\('.$sm[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($sm[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@sm)),
  ],
  center=>0,
  align=>'ll',
  midrules=>1,
) @]*
END

BEGIN_PGML
[$TBL]**
----
END_PGML


##########
##########
##########
##########


BEGIN_PGML
Now start over, with 'digits' tolerance, with 'tolerance' 3, and 'tolTruncation' _false_.  
END_PGML

Context("Numeric")->flags->set(tolType=>'digits',tolTruncation=>0);
# Note the default tolerance of 0.001 is converted to tolerance of 3

#Start over to avoid context having information about any previous comparisons 

# a value, then the expecations for comparison to 0th value
@pi = (
[Real(pi),'T','T'],
[Real(3.14),'T','T'],
[Real(3.15),'F','F'],
[Real(3.13),'F','F'],
[Real(3.141),'T','F'],
[Real(3.142),'T','T'],
[Real(3.14159),'T','T'],
[Real(3.14149),'F','F'],
);

@lz = (
[Real(0.1055),'T','T'],
[Real(0.105),'T','F'],
[Real(0.106),'T','T'],
);

@nlz = (
[Real(-0.1055),'T','T'],
[Real(-0.105),'T','F'],
[Real(-0.106),'T','T'],
);

@lg = (
[Real(1234e20),'T','T'],
[Real(123000000000000000000000),'T','T'],
[Real(124000000000000000000000),'F','F'],
);

@sm = (
[Real(0),'T','T'],
[Real(10**(-14)),'F','F'],
[Real(10**(-15)),'T','T'],
);


$TBL = <<'END';
[@ DataTable(
  [
    ['','Equal?','Expecation allowing truncation','Expecation not allowing truncation'],
    (map{['\('.$pi[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($pi[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@pi)),
    (map{['\('.$lz[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($lz[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@lz)),
    (map{['\('.$nlz[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($nlz[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@nlz)),
    (map{['\('.$lg[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($lg[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@lg)),
    (map{['\('.$sm[0]->[0].'\stackrel{?}{=}'.$_->[0].'\)',($sm[0]->[0]==$_->[0])?'T':'F',$_->[1],$_->[2]]}(@sm)),
  ],
  center=>0,
  align=>'ll',
  midrules=>1,
) @]*
END

BEGIN_PGML
[$TBL]**
----
END_PGML




ENDDOCUMENT();
```

drgrice1 commented:
All of the tests are checking out for me now.

Is it concerning that changing flags does not take effect without re-computing?  I could see in a problem with a custom grader an instructor could try changing flags to make both comparisons.

drdrew42 commented:
Objects inherit the current context in which they are defined. This should have occurred to me before, but the differences between the two can be seen by using `Context()->flags->set(tolType=>'digits',tolTruncation=>1);` 
instead of `Context("Numeric")->flags->set(tolType=>'digits',tolTruncation=>1);` in the earlier code.

Calling `Context()` 'pulls up' the existing context for modification.
Calling `Context("Numeric")` creates a new context, leaving the already declared MO behind in the first context.

drgrice1 commented:
@drdrew42: I understand the difference in the Context calls.  What I meant is similar to what is done when using the bizzaro comparison type things.  When you change flags on the context, you change the behavior of the operations in the context.

drdrew42 commented:
The behavior you're describing (with bizarroArithmetic) relies on an empty argument to Context, and would be expected to break if flags were set/changed on a freshly-declared context instead. 
What we are seeing here is the same. 
The re-definition of the Real objects in the latest testing code is unnecessary if, instead of declaring a *new* context prior to `tolTruncation=>0`, we instead change the flags on the existing context.
Maybe we're just talking past each other here, or else I'm not understanding the concern.

drgrice1 commented:
I see now. I didn't look closely at where the first Context call occurred in that middle version.

dpvc commented:
I will try to take a look at this this weekend.

dpvc commented:
OK, I've taken a look, and I have some comments to make, but I won't be able to write them up until tomorrow, most likely.

dpvc commented:
Sorry to be late to the party, and sorry for the length of this comment.  There are some subtleties that that require a bit of description, however.

Although you say you you don't think 3.14177 should count as *pi*, you don't really say what the digits check it actually *supposed* to do, so it is not easy to determine if it actually does that or not.  So first I had to decide what I thought you were trying to accomplish.  Based on the algorithm, this is what I think the intent is.  Let *n* be the number of digits specified by `tolerance` and *N* the number from `tolDigits`.  Then if one of the numbers has no more than *d* digits, where *n* ≤ *d* ≤ *N*, then the two are equal if they are equal when rounded (or truncated) to *d* digits, otherwise they are equal if they agree to *N* digits.  I hope that is correct, because the rest of this comment is based on that understanding.

I would guess that the reasoning for this would be that if a student needs to enter at least 3 digits, but enters more, then you want those digits to be correct as well.  That makes sense for numbers entered directly by students, but I don't think it will work well for answers that result from computations, as round-off errors can interfere with the "has no more than *d* digits" part.  For example, with *pi* to 3 digits, consider the case where the student enters `3.14 + 100(1.6-1.5) - 10`.  Mathematically, 1.6 - 1.5 is .1, and so 100 (1.6 - 1.5) is 10, and 100(1.6 - 1.5) is 0, so the student has entered 3.14 exactly.  But the current implementation will mark this as incorrect.  (It is interesting to figure out where the round-off error is coming from, and how it affects the result of the algorithm.  I some of it away below.)

Because computations can cause round-off errors like this, determining whether the student answer has exactly *d* digits is always going to be fragile, and will fail in some cases.  It may be reasonable to do this with numbers typed directly by the student, but not with computed values.  This is inherent in trying to represent base-10 numbers using a fixed number of binary digits, as terminating decimals in base 10 may be repeating in binary.  E.g., .1 is a repeating "decimal" in binary, so can't be represented accurately as a finite-precision binary number.  That is at the heart of the example above.  That means you can't really tell whether you have a finite number of decimal digits, because the internal representation of .1 is never completely accurate, and you can't tell if what you have is actually supposed to be .1 or is just very close to it.  Moreover, the error in representing .1 may propagate into higher digits during computations, and so will make things that should have a small number of digits end up seeming to have many more (as in the case above, where you end up with 3.14000000000001 rather than 3.14 exactly).  Trying to rely on having all zeros from some point on means you are going to fail for some computations.

While you may say that rounded answers are only likely to be entered as directly typed-in answers, not computed answers, so this kind of issue doesn't matter, there is another consequence that needs to be considered.  For computed values (that are not rounded), this effectively means you have to be correct to `tolDigits` digits, so you have upped the precision requirements to 6 digits from the usual roughly 4 digits.  That is a significant change, and one that would not be in line with most of the documentation that says the answers need to be correct to 4 digits or so.  Also, the conversion of a tolerance of .001 to 3 digits reduces the requirement in most cases.  For example, with the standard relative tolerances, `Real(1/9)` requires `.1111` to match, not `.111` and `Real(8/9)` requires `.8888` not `.888` or `.889`.  So you seem to be increasing the requirements in one direction and decreasing them in another.

Given those parameter, however, another potential issue is that the check is symmetric, meaning that if either one is a rounding (or truncation) of the other, then they are considered to be equal.  So if the correct answer is 3.14 exactly, then a student could enter 3.141 and it would be marked correct.  That doesn't seem to be in keeping with the idea that if the student enters extra digits, they should be correct.  I think the check should not be symmetric, but should only allow the student's answer to be rounded or truncated, not the correct one.  Note that we already have non-symmetric comparison with the current `tolType => "relative"`, which is based on the absolute error relative to the correct answer:  (correct - student) / |correct| < tolerance.  This is asymmetric, so there is no reason the test must be symmetric.  (The MathObject answer checkers always use the first value as the correct one and the second as the student one when testing equality so as to use the correct relative-error test.  That is also why you should always use the correct answer first when comparing with a student answer in custom checkers as well.)

So those are reasons why I think the premise of the digits check is problematic, though all of these can be addressed.  In addition, however, I think there are some implementation issues as well.  For example, I would expect that if the correct answer is 75.1249, then 75.125 should be marked as correct (with `tolerance` of 3 and `tolDigits` of 6), since it is the correct answer rounded to 5 digits. Ironically, it is marked false when only rounding it allowed, but true when truncation is also allowed.  This is because the algorithm requires the rounding to be equal for every number of digits between *n* and *d*, and in this case, that would require 75.12 == 75.13 before it tests 75.125 == 75.125, and so the test fails.

Similarly, I would expect that if the correct answer is 1.235, then 1.24 should be marked correct as the rounding to two digits, but this is marked false regardless of the truncation setting.  That is because it turns out that `int(1.235*10**2 + 0.5)/10**2` ends up being 1.23 rather than 1.24, so the rounding doesn't work as expected in some cases.  This is the problem I mentioned above cropping up its ugly head again, where 1.235 is not accurately represented in a finite number of binary digits.

Another example uses 1.200005 as the correct answer (which does round to 6 digits as 1.20001).  When compared to 1.200004999999999, it is marked incorrect, but 1.2000049999999999 (one more 9) is marked correct.  That means you are requiring accuracy to 10^-15, which seems a bit excessive if you are allowing rounding to three digits.  A similar issue arrises with 1.203456, where 1.20 should be a correct answer (or even just 1.2).  But what about 1.19999999999999?  This is marked incorrect, even though it is off by only 10^-14, while 1.199999999999999 is marked correct.  Computed answers can easily be off by these amounts, so this could cause very accurate answers to be marked wrong.  Since 1.19999999999999 will display as 1.2, a correct answer, this can lead to confusion for students.

For these reasons, I would be hesitant to use `digits` as the default `tolType`.

There is one other esoteric issue that I noticed as well, which will not show up with student answers, but I thought I'd mention it anyway.  It turns out that `Real(1234E+20)` and `Real("123000000000000000000000")` are not equal, while `Real(1234E+20)` and `Real(23000000000000000000000)` *are* equal.  It took me a while to figure out why.  The reason is subtle:  The `Real()` constructor checks to see if its argument looks like a number (by doing a regex match against the number pattern from the context), and if so, it sets the MathObject's value to that argument (otherwise it essentially calls Compute() to compute the value as a formula, so that `Real("1+2")` will do what is expected).  That means that for `Real("123000000000000000000000")` the internal value is the string `123000000000000000000000` while for `Real(123000000000000000000000)` it is the actual Perl number represented by `123000000000000000000000`.  Normally that doesn't make a different, since Perl will convert a string to a number when it needs to, so `"12" + 3` does equal 15.  But the algorithm uses

```
last if ($a eq $rounda or $b eq $roundb);
```

as a termination condition, and so this is a string comparison, not a numeric one.  When `$a` is the string `"123000000000000000000000"`, `$rounda` will be the number `123000000000000000000000`, which stringifies as `1.23E+23`, so these aren't equal.  I'm assuming that the string check is supposed to cause things like 1.199999999999999 to compare as 1.2, which usually works, but not in this case.  One solution to this would be to alter the `new` method for the `Value::Real` object to force its internal value to be a Perl real (say by adding 0 to it).

Here is a modified test file that includes the original tests, but tests that illustrate the issues I mentioned above.  I've modified the table slightly so that it lists the test together with the expected values for with and without truncation, where the expected result is colored green if the test produces that value, and red if not.  So a red "T" means that the test should be true, but is returning false.

``` perl
DOCUMENT();

loadMacros(
  "PGstandard.pl",
  "MathObjects.pl",  
  "PGML.pl",
  "niceTables.pl",
  "PGcourse.pl",
);

TEXT(beginproblem());

########################################################

sub T {1 == 1}
sub F {1 == 0}

#
#  The tests to perform
#

$tests = [
  {
    correct => pi,
    check => [
      [3.14, T, T],
      [3.15, F, F],
      [3.13, F, F],
      [3.141, T, F],
      [3.142, T, T],
      [3.14159, T, T],
      [3.14149, F, F],
      [3.1415911, T, T],
    ]
  }, {
    correct => 0.1055,
    check => [
      [0.105, T, F],
      [0.106, T, T],
    ]
  }, {
    correct => -0.1055,
    check => [
      [-0.105, T, F],
      [-0.106, T, T],
    ]
  }, {
    correct => 1234E+20,
    check => [
      ['123000000000000000000000', T, T],
      ['124000000000000000000000', F, F],
    ]
  }, {
    correct => 0,
    check => [
      ['1E-14', F, F],
      ['1E-15', T, T],
    ]
  }, {
    extra => 1,
    noInitial => 1,
    correct => pi,
    check => [
      ['3.14 + 100 (1.6 - 1.5) - 10', T, T]
    ]
  }, {
    noInitial => 1,
    correct => 3.14,
    check => [
      [3.141, F, F],
    ]
  }, {
    noInitial => 1,
    correct => 75.1249,
    check => [
      [75.125, T, T],
    ]
  }, {
    noInitial => 1,
    correct => 1.235,
    check => [
      [1.24, T, T],
    ]
  }, {
    noInitial => 1,
    correct => 1.24,
    check => [
      [1.235, F, F],
    ]
  }, {
    correct => 1.200005,
    check => [
      [1.20001, T, T],
      [1.20000499999999, T, T],
      ['1.200005 - 1E-15', T, T],
    ]
  }, {
    noInitial => 1,
    correct => 1.203456,
    check => [
      [1.2, T, T],
      [1.203, T, T],
      [1.2035, T, T],
      [1.2034, T, F],
      [1.19999, F, F],
      [1.19999999999999, T, T],
    ]
  }, {
    noInitial => 1,
    correct => 1234E+20,
    check => [
      [Real('123000000000000000000000'), T, T],
    ]
  }
];

########################################################
#
#  Compute the test table
#

$arrow = PGML::LaTeX('\(\Longrightarrow\)');

#
#  A colored T or F
#
sub expected {
  my ($e, $r) = @_;
  my $color = ($e == $r ? 'green' : 'red');
  my $c = ($e ? 'T' : 'F');
  return qq{<span style="color: $color">$c</span>};
}

#
#  One row of the table
#
sub testRow {
  my ($c, $s, $r, $t) = @_;
  $s = Compute($s) unless Value::isValue($s);
  Context()->flags->set(tolTruncation => 1);
  $v1 = ($c == $s);
  Context()->flags->set(tolTruncation => 0);
  $v2 = ($c == $s);
  $s = $_[1];  $s =~ s/1E(.*)/10^{$1}/;
  return [PGML::LaTeX("\($c = $s\)"), $arrow, expected($r, $v1), expected($t, $v2)];
}

#
#  Compute the rows for one group of tests
#
sub testGroup {
  my ($c, $tests, $noInitial) = @_;
  $c = Real($c);
  return (
    ($noInitial ? () : testRow($c, $c, T, T)),
    map {testRow($c, @$_)} (@$tests)
  );
}

#
#  Compute all the tests in the list and form the table
#
sub runTests {
  my $type = shift;
  my $tests = shift;
  my @rows = (["Test ($type)", '', 'Trunc', 'No Trunc']);
  foreach my $test (@$tests) {
    push(@rows, ['', '', '', '']) if $test->{extra};
    push(@rows, testGroup($test->{correct}, $test->{check}, $test->{noInitial}));
    push(@rows, ['', '', '', '']);
  }
  pop(@rows);
  return DataTable(
    [@rows],
    midrules => 1,
    align => 'llcc'
  );
}

Context()->texStrings;   #  Make MathObjects insert their TeX forms


#####################################################
#
#  Set up the context
#

Context("Numeric")->flags->set(
  tolType => 'digits',
  tolDigits => 6,
);
Context()->{format}{number} = "%.8g";   # so we can see past 6 digits.
# Note the default tolerance of 0.001 is converted to tolerance of 3

#
#  Run the tests
#
TEXT(runTests("digits", $tests));

ENDDOCUMENT();
```

I will be making a pull request that addresses the concerns expressed above, and that eliminates the red entries above.

dpvc commented:
Sorry if that comment sounded harsh.  I don't mean it that way, but don't know how else to point out the issues that I saw.  I have spent more time than I should have on it, and didn't have much time to polish things up.  So if I offended, I'm sorry.

Alex-Jordan commented:
Not at all, and thank you for reviewing this. The detail of your comments is much appreciated, and I'll digest this over the coming week. I'm sorry that you had to do forensics to figure out the intent. I thought the PR intro covered that, and I'd write wiki documentation later. But it would have been better to comment the code more formally.

We've had a long-time complaint here from local faculty that set me off to find a solution. Plenty of faculty here (but not me) use Desmos with their classes. Desmos has a habit of only presenting three significant digits to numbers on graphs. But only sometimes, and I'm unsure what triggers Desmos to show 3 versus 4. So with the status quo, every once in a while a student has "the correct" answer—the infallible Desmos said so. But WeBWorK rejects it, because it is outside of the 0.1% relative tolerance. And Desmos does not give users an option to show more precision like GeoGebra does. [Even though that feature has been requested many times.]

I've tried telling colleagues that this is a Desmos problem, not a WW problem, but that doesn't win people over.

Reducing relative tolerance globally to 1% is not an option, because that is such a wide window of tolerance in other cases. 

Visiting all the problems in our library and moving each one to an absolute tolerance is too big of a project and unsustainable, not to mention the complications for when randomization can make answers have different orders of magnitude.

So I set out to think like my human colleagues think when they have a student paper in front of them with a decimal answer. As long as decimals aren't explicitly banned, they are OK with decimals, but not entirely OK when the digits diverge from what they should be. At minimum, there might be a comment about rounding error. What I came up with is exactly what you describe with n, d, and N. The N is only needed because the system could never keep checking forever, and also it partly mitigates the machine rounding issues. I called it `digits` instead of `sigfig` since it doesn't account for trailing zeros and it has its own kind of tolerance window, which maybe a true `sigfig` would not have.

Thanks again for these insights. They may be insurmountable, but I'll chew on them.

dpvc commented:
I've made a PR at Alex-Jordan/pg#1 that implements an alternative approach.  This resolves most of the issues I mention above, but may not be exactly what Alex had in mind (especially in how lower-order digits are ignored).  I suggest a possible adjustment in that PR to address that as well, but have run out of time to work on it further.   See what folks think of that approach.  (I have left both version in the file so that they could be compared).

dpvc commented:
@Alex-Jordan, your issue with Desmos makes sense of your choice for 3 digits.  Thanks for the explanation.  Sounds like it is a bit of a hassle.  

I agree that this is not a significant-figures checker, and doesn't intend to be.  That can't be done from the number itself, but would require looking at the original student input, and I don't think you could allow computation to be involved for that.  I worked something out once for that and will have to look it up to see if it is any good...

pstaabp reviewed (approved):
Looks as advertised.

taniwallach commented:
I quickly read the thread and have not tested the code. I would rather test once we decide on the proposal below.

I think there are certainly settings where such a tolerance type can be very helpful, even if the implementation to the underlying floating point precision issues are technically imperfect in some cases. As  this is a bleeding-edge new feature, accepting it, even if some future changes will not be fully backwards compatible is probably beneficial to the early adopters and many other potential users.

How about this: 

1. Merging @dvpc's PR which adds `digits-2` into Alex's branch.
2. Rename @Alex-Jordan 's original `digits` as `digits-1`.
3. Mark both as **provisional** in the documentation (in the code and on the Wiki) and warn about the potential for future changes to possible break some problems using the initial version.
4. Then merge the PR and add both to the system.

In the documentation refer to the two options, that they are still provisional, and some of the reasons to prefer one over the other. 

Anyone who uses this can then test their new problems with both options and use the one which better meets their expectations.

If future changes are likely to break something, we could decide to provide a `digits-1-original-draft` option keeping the old code alive, at least for a while.

Alex-Jordan commented:
Just dropping a note that I've had significant local issues to work on.
Before anything is merged, I would like to try Davide's version and think
through the differences. But it may be next week now. As Tani suggests, if
it's a new feature that is marked as provisional, maybe it's not so bad if
it's merged at the 11th hour.

On Wed, Mar 10, 2021 at 6:52 AM Nathan Wallach <notifications@github.com>
wrote:

> I quickly read the thread and have not tested the code. I would rather
> test once we decide on the proposal below.
>
> I think there are certainly settings where such a tolerance type can be
> very helpful, even if the implementation to the underlying floating point
> precision issues are technically imperfect in some cases. As this is a
> bleeding-edge new feature, accepting it, even if some future changes will
> not be fully backwards compatible is probably beneficial to the early
> adopters and many other potential users.
>
> How about this:
>
>    1. Merging @dvpc <https://github.com/dvpc>'s PR which adds digits-2
>    into Alex's branch.
>    2. Rename @Alex-Jordan <https://github.com/Alex-Jordan> 's original
>    digits as digits-1.
>    3. Mark both as *provisional* in the documentation (in the code and on
>    the Wiki) and warn about the potential for future changes to possible break
>    some problems using the initial version.
>    4. Then merge the PR and add both to the system.
>
> In the documentation refer to the two options, that they are still
> provisional, and some of the reasons to prefer one over the other.
>
> Anyone who uses this can then test their new problems with both options
> and use the one which better meets their expectations.
>
> If future changes are likely to break something, we could decide to
> provide a digits-1-original-draft option keeping the old code alive, at
> least for a while.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwebwork/pg/pull/441#issuecomment-795542501>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEDOAD36UR2IZMHFCBJSULTC6BTPANCNFSM4JTFSWPQ>
> .
>


-- 
Alex Jordan
Mathematics Instructor
Portland Community College

taniwallach commented:
@Alex-Jordan wrote:
> maybe it's not so bad if it's merged at the 11th hour.

It is also something we could consider merging after making a release candidate branch. No one expects the new context as part of the base PG yet, and should not be able to break something else when merged. If that is acceptable to the release team, the urgency of finishing this is far lower and will reduce stress for @Alex-Jordan .

dpvc commented:
I'm not sure it is necessary to have two version of essentially the same thing.  If you put out both, people will start using them, and then it will be harder to remove one, provisional or not.  The difference are not that great, so I think it would be best to just pick one.

Alex-Jordan commented:
Question for the group: where is the right place to document this? One option is right before `sub compare` in `Value/Real.pm`, I could document all three tolerance types and also periodic reals.

Right now I think documentation only exists in the wiki. Should I just leave it out of the code and put something in the wiki soon?

dpvc commented:
Nothing is POD documented in the MathObejct `.pm` files, and little else use in comments.  So I'd go with the wiki for now.

Alex-Jordan commented:
I took Davide's version, made some small changes, rebased from develop (since I forgot I was working on a branch where this had been merged into develop) and force pushed. I think this is ready for testing and/or merging.

Here is Davide's [test problem](http://spot.pcc.edu/~ajordan/digitsTestProblem.pg), modified a little for the changes I made, and also now it compares `digits` to `relative` (0.1%) instead of with `digits-2`.
